src/main/java/com/acoustic/SpringPolandSalaryCalculator/controller/SalaryCalculatorController.java

```diff
@@ -0,0 +1,124 @@
+
+
+    @GetMapping("/getJobTitles/{departmentName}")
+    public List<String> getJobTitles(
+            @PathVariable
+            String departmentName) {
+        return this.jobCategoriesConfigurationProperties.getJobDepartmentAndTitles().get(departmentName);
+    }
+
+    @GetMapping("/getJobDepartments")
+    public Set<String> getDepartmentName() {
+        return this.jobCategoriesConfigurationProperties.getJobDepartmentAndTitles().keySet();
+    }
+
+
+    @PostMapping("/calculator/{grossMonthlySalary}")
+    public Map<String, BigDecimal> calculateSalary(
+            @PathVariable
+            @Min(value = MINIMUM_GROSS,
+                    message = "Must be Greater than or equal to 2000.00") @NotNull BigDecimal grossMonthlySalary,
+            @RequestParam(required = false)
+            String departmentName,
+            @RequestParam(required = false)
+            Integer jobTitleId) {
+
+        var response = responseWithoutStatistics(grossMonthlySalary);
+        if (departmentName == null || jobTitleId == null) {
+            return response;
+        }
+        return getAverage(grossMonthlySalary, departmentName, jobTitleId, response);
+
+    }
+
+    private Map<String, BigDecimal> responseWithoutStatistics(BigDecimal grossMonthlySalary) {
+        Map<String, BigDecimal> response = new LinkedHashMap<>();
+        salaryCalculatorServices.sort(Comparator.comparingInt(SalaryCalculatorService::getCalculationOrder));
+        BigDecimal grossMonthlySalaryMinusTaxes = grossMonthlySalary;
+        BigDecimal tax = null;
+        final int taxesOrder = 3;
+        final int netOrder = 4;
+        for (var service : salaryCalculatorServices) {
+            if (service.getCalculationOrder() > 5) {
+                response.put(service.getDescription(), service.apply(grossMonthlySalary).setScale(2, RoundingMode.HALF_EVEN));
+            } else {
+                response.put(service.getDescription(), service.apply(grossMonthlySalaryMinusTaxes).setScale(2, RoundingMode.HALF_EVEN));
+
+                if (service.getCalculationOrder() == taxesOrder) {
+                    response.put(service.getDescription(), tax = service.apply(grossMonthlySalary));
+                    continue;
+                }
+                if (service.getCalculationOrder() == netOrder) {
+                    response.put(service.getDescription(), service.apply(grossMonthlySalaryMinusTaxes = grossMonthlySalaryMinusTaxes.subtract(tax)));
+                    continue;
+                }
+                grossMonthlySalaryMinusTaxes = grossMonthlySalaryMinusTaxes.subtract(service.apply(grossMonthlySalaryMinusTaxes));
+            }
+        }
+        return response;
+    }
+
+    private Map<String, BigDecimal> getAverage(
+            BigDecimal grossMonthlySalary, String departmentName, int jobTitleId, Map<String, BigDecimal> response) {
+        if (this.jobCategoriesConfigurationProperties.getJobDepartmentAndTitles()
```

Comment:
What do you think about doing such validation on the first function (in this case `calculateSalary`) and returning an error instead of in the middle of the code? Same for `statistic`?

Reply:
Done